server.c

```diff
@@ -0,0 +1,292 @@
+    struct srv_context_t *context;
+    int socket_fd;
+};
+
+static int
+send_task(const int client_sfd, struct task_t *task, bool *result)
+{
+    int status;
+
+    status = send(client_sfd, (char *) task, sizeof(struct task_t), 0);
+    if (status <= 0) return -1;
+
+    int size;
+    status = recv(client_sfd, &size, sizeof(size), 0);
+    if (status <= 0) return -1;
+
+    if (size != 0)
+        recv(client_sfd, task->password, size, 0);
+
+    *result = (size != 0);
+    return 0;
+}
+
+static void *
+serve_client(void *arg)
+{
+    struct params_t *params = (struct params_t *) arg;
+    struct srv_context_t *context = (struct srv_context_t *) params->context;
+    int client_sfd = params->socket_fd;
+
+    while (true)
+    {
+        struct task_t task;
+        queue_pop(&context->queue, &task);
+
+        task.to = task.from;
+        task.from = 0;
+        bool found = false;
+        int status = send_task(client_sfd, &task, &found);
+        if (status == -1)
+        {
+            queue_push(&context->queue, &task);
+            set_remove_sock(&context->set, client_sfd);
+            close(client_sfd);
+            break;
+        }
+        if (found)
+        {
+            memcpy(context->password, task.password, sizeof(task.password));
+            context->found = true;
+            context->done = true;
+        }
+
+        pthread_mutex_lock(&context->tasks_mutex);
+        --context->tasks_running;
+        pthread_mutex_unlock(&context->tasks_mutex);
+
+        if (context->tasks_running == 0 || context->found)
+            pthread_cond_signal(&context->tasks_cond);
+    }
+
+    return NULL;
+}
+
+static bool
+srv_password_handler(void *context, struct task_t *task)
+{
+    struct srv_context_t *ctx = (struct srv_context_t *) context;
+
+    pthread_mutex_lock(&ctx->tasks_mutex);
+    ++ctx->tasks_running;
+    pthread_mutex_unlock(&ctx->tasks_mutex);
+
+    queue_push(&ctx->queue, task);
+    return (ctx->password[0] != 0);
+}
+
+static void *
+srv_producer(void *arg)
+{
+    struct srv_context_t *context = (struct srv_context_t *) arg;
+
+    struct task_t task;
+    task.from = 2;
+    task.to = context->config->length;
+
+    process_task(&task, context->config, context, srv_password_handler);
+    context->done = true;
+    return NULL;
+}
+
+static void *
+srv_server(void *arg)
+{
+    struct params_t *params = (struct params_t *) arg;
+    struct srv_context_t *context = (struct srv_context_t *) params->context;
+    int server_sfd = params->socket_fd;
+
+    while (true)
+    {
+        // Printing to stderr to be able to check output in tests
+        fprintf(stderr, "Waiting for new connections...\n");
+        int client_socket = accept(server_sfd, NULL, NULL);
+        if (client_socket == -1)
+            handle_error("accept");
+        fprintf(stderr, "Got new connection...\n");
+
+        struct node_t *node = set_take_last(&context->set);
+        pthread_create(&node->thread_id, NULL, serve_client,
```

Comment:
Here is a race condition

Reply:
&(struct params_t) { context, client_socket } is a temporary variable on the stack. Iteration for the next client will rewrite content of this variable. If thread have not started yet, it will get inconsistent data